### PR TITLE
(PUP-10648) simplify code and remove exception usage

### PR DESCRIPTION
**Pull request by ciprianbadescu**
follow-up from https://github.com/puppetlabs/puppet/pull/8354#discussion_r504258393

### CODE REVIEW

**Comment by ciprianbadescu** [comment]
Diff with the code before initial change:
```
@@ -226,7 +226,18 @@ class Puppet::Configurer
         if options[:catalog].nil? && do_failover
           server, port = find_functional_server
           if server.nil?
-            raise Puppet::Error, _("Could not select a functional puppet master from server_list: '%{server_list}'") % { server_list: Puppet.settings.value(:server_list, Puppet[:environment].to_sym, true) }
+            detail = _("Could not select a functional puppet server from server_list: '%{server_list}'") % { server_list: Puppet.settings.value(:server_list, Puppet[:environment].to_sym, true) }
+            if Puppet[:usecacheonfailure]
+              options[:pluginsync] = false
+              @running_failure = true
+              if server.nil?
+                server = Puppet[:server_list].first[0]
+                port = Puppet[:server_list].first[1] || Puppet[:masterport]
+              end
+              Puppet.err(detail)
+            else
+              raise Puppet::Error, detail
+            end
           else
             #TRANSLATORS 'server_list' is the name of a setting and should not be translated
             Puppet.debug _("Selected puppet server from the `server_list` setting: %{server}:%{port}") % { server: server, port: port }
```

**Comment by joshcooper** [comment]
jenkins please test this with servertests

**Review by joshcooper — APPROVED** [review]

File: lib/puppet/configurer.rb [hunk 1]
```diff
@@ -225,26 +225,23 @@ def run(options = {})
...
-              Puppet.debug _("Selected puppet server from the `server_list` setting: %{server}:%{port}") % { server: server, port: port }
-              report.master_used = "#{server}:#{port}"
-            end
-          rescue Puppet::Error => detail
+          if server.nil?
+            detail = _("Could not select a functional puppet server from server_list: '%{server_list}'") % { server_list: Puppet.settings.value(:server_list, Puppet[:environment].to_sym, true) }
             if Puppet[:usecacheonfailure]
               options[:pluginsync] = false
               @running_failure = true
               if server.nil?
```
Comment (joshcooper) [inline lib/puppet/configurer.rb:233]: `server.nil?` is already checked above on line 228

**Comment by joshcooper** [comment]
jenkins please test this with servertests

**Review by joshcooper — APPROVED** [review]

**Pull request merged**